tests/testthat/test-dataframe-s3-base.R

```diff
@@ -25,3 +25,122 @@ test_that("tail() works", {
     ignore_attr = TRUE
   )
 })
+
+test_that("`[` operator works to subset columns only", {
+  test <- pl$DataFrame(a = 1:3, b = 4:6, c = 7:9)
+
+  ### Indices
+  expect_identical(test[1], pl$DataFrame(a = 1:3))
+  expect_warning(
+    expect_identical(test[1, drop = TRUE], pl$DataFrame(a = 1:3)),
+    "`drop` argument ignored for subsetting a frame",
+  )
+  expect_identical(test[, 1, drop = TRUE], list(a = pl$Series("a", 1:3)))
+  expect_identical(test[1:2], pl$DataFrame(a = 1:3, b = 4:6))
+  expect_identical(test[, 1:2, drop = TRUE], pl$DataFrame(a = 1:3, b = 4:6))
+  expect_identical(test[, -2:-1], pl$DataFrame(c = 7:9))
+  expect_snapshot(test[, 10:12], error = TRUE)
+  expect_snapshot(test[, -2:1], error = TRUE)
+
+  ### Column names
+  expect_identical(test["a"], pl$DataFrame(a = 1:3))
+  expect_warning(
+    expect_identical(test["a", drop = TRUE], pl$DataFrame(a = 1:3)),
+    "`drop` argument ignored for subsetting a frame",
+  )
+  expect_identical(test[, "a", drop = TRUE], list(a = pl$Series("a", 1:3)))
+  expect_identical(test[c("a", "b")], pl$DataFrame(a = 1:3, b = 4:6))
+  expect_identical(test[, c("a", "b"), drop = TRUE], pl$DataFrame(a = 1:3, b = 4:6))
+  expect_error(
+    test[c("a", "foo")],
+    "not found: foo"
+  )
+
+  ### Logical values
+  expect_identical(test[TRUE], test)
+  expect_identical(test[c(TRUE, TRUE, FALSE)], pl$DataFrame(a = 1:3, b = 4:6))
+  expect_error(
+    test[c(TRUE, FALSE)],
+    "must be size 1 or 3, not 2"
+  )
+  expect_identical(test[, TRUE, drop = TRUE], test)
+  expect_identical(
+    test[, c(TRUE, FALSE, FALSE), drop = TRUE],
+    list(a = pl$Series("a", 1:3))
+  )
+
+  ### Empty args
+  expect_identical(test[], test)
+  expect_identical(test[,], test)
+})
+
+test_that("`[` operator works to subset rows only", {
+  test <- pl$DataFrame(a = 1:3, b = 4:6, c = 7:9)
+
+  ### Indices
+  expect_identical(
+    test[1, ],
+    pl$DataFrame(a = 1L, b = 4L, c = 7L)
+  )
+  expect_identical(
+    test[1:2, ],
+    pl$DataFrame(a = 1:2, b = 4:5, c = 7:8)
+  )
+  # TODO
+  # expect_identical(
+  #   test[c(1, 10), ],
+  #   pl$DataFrame(a = c(1L, NA_integer_), b = c(4L, NA_integer_), c = c(7L, NA_integer_))
+  # )
```

Comment:
I didn't find a good way to include NA-filled rows for out of bound indices. There a few TODOs here related to that.